tsdb/importer/importer.go

```diff
@@ -0,0 +1,602 @@
+		var err error
+		advance := 0
+		lineIndex := 0
+		lines := strings.Split(string(data), "\n")
+		parser := textparse.New(data, contentType)
+		for {
+			var et textparse.Entry
+			if et, err = parser.Next(); err != nil {
+				if !atEOF && et == textparse.EntryInvalid && !currentEntryInvalid {
+					currentEntryInvalid = true
+					// Fetch more data for the scanner to see if the entry is really invalid.
+					return 0, nil, nil
+				}
+				if err == io.EOF {
+					err = nil
+				}
+				return 0, nil, err
+			}
+
+			// Reset invalid entry flag as it meant that we just had to get more data
+			// into the scanner.
+			if currentEntryInvalid {
+				currentEntryInvalid = false
+			}
+
+			// Add 1 to account for newline character.
+			lineLength := len(lines[lineIndex]) + 1
+			lineIndex++
+
+			if et != textparse.EntrySeries {
+				advance = advance + lineLength
+				continue
+			}
+
+			_, ctime, cvalue := parser.Series()
+			if ctime == nil {
+				return 0, nil, NoTimestampError
+			}
+			// OpenMetrics parser multiples times by 1000 - undoing that.
+			ctimeCorrected := *ctime / 1000
+
+			var clabels labels.Labels
+			_ = parser.Metric(&clabels)
+
+			sample := tsdb.MetricSample{
+				Timestamp: ctimeCorrected,
+				Value:     cvalue,
+				Labels:    labels.FromMap(clabels.Map()),
+			}
+
+			buf.Reset()
+			enc := gob.NewEncoder(buf)
+			if err = enc.Encode(sample); err != nil {
+				return 0, nil, err
+			}
+
+			advance += lineLength
+			return advance, buf.Bytes(), nil
+		}
+	}
+}
+
+// makeRange returns a series of block times between start and stop,
+// with step divisions.
+func makeRange(start, stop, step int64) []blockTimestampPair {
+	if step <= 0 || stop < start {
+		return []blockTimestampPair{}
+	}
+	r := make([]blockTimestampPair, 0)
+	// In case we only have samples with the same timestamp.
+	if start == stop {
+		r = append(r, blockTimestampPair{
+			start: start,
+			end:   stop + 1,
+		})
+		return r
+	}
+	for s := start; s < stop; s += step {
+		pair := blockTimestampPair{
+			start: s,
+		}
+		if (s + step) >= stop {
+			pair.end = stop + 1
+		} else {
+			pair.end = s + step
+		}
+		r = append(r, pair)
+	}
+	return r
+}
+
+// mergeBlocks looks for blocks that have overlapping time intervals, and compacts them.
+func mergeBlocks(dest string, metas []*newBlockMeta, logger log.Logger) error {
+	for _, m := range metas {
+		// If no children, there's nothing to merge.
+		if len(m.children) == 0 {
+			continue
+		}
+		if err := compactMeta(dest, m, logger); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// binBlocks groups blocks that have fully intersecting intervals of the given durations.
+// It iterates through the given durations, creating progressive blocks, and once it reaches
+// the last given duration, it divvies all the remaining samples into blocks of that duration.
+// Input blocks are assumed sorted by time.
+func binBlocks(blocks []*newBlockMeta, durations []int64) [][]*newBlockMeta {
```

Comment:
What `bin` means here? Can we have more verbose function name? (:

Reply:
Yeah, that function basically creates bins, or tranches, of the given samples corresponding to the provided durations. I thought bin was an accurate term for it

Reply:
bin like... garbage bin? :thinking: